python/lsst/ts/observatory/control/base_group.py

```diff
@@ -0,0 +1,541 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+import enum
+import types
+import asyncio
+import logging
+import traceback
+
+from lsst.ts import salobj
+
+__all__ = ["BaseUsages", "BaseGroup"]
+
+
+class BaseUsages(enum.IntEnum):
+    """Enumaration defining base class usages. When subclassing `BaseGroup`
+    it is recommended to define new enumerations.
+
+    Notes
+    -----
+
+    Usages definition:
+
+    All: Enable all possible operations defined in the class. This is
+    different than adding all topics from all components. It will only
+    add those needed to run the methods defined in the class.
+
+    StateTransition: Enable summary state transition.
+
+    MonitorState: Enable monitoring of summary state.
+
+    MonitorHeartBeat: Enable monitoring of heartbeat.
+    """
+
+    All = 0
+    StateTransition = 1
+    MonitorState = 1 << 1
+    MonitorHeartBeat = 1 << 2
+
+
+class BaseGroup:
+    """High-level abstraction of a collection of `salobj.Remote`.
+
+    A `salobj.Remote` provides a communication abstraction for Commandable
+    SAL Componentes (CSCs). CSCs represents entities in the LSST observatory
+    system that performs actions such as controlling/interfacing with hardware,
+    logical operations and/and coordination of other CSCs. In general CSCs
+    operate in logical groups to achieve certain kinds of operation; like
+    slewing the telescope and taking images.
+
+    Parameters
+    ----------
+    components : `list` [`str`]
+        A list of strings with the names of the components that are part
+        of the collection.
+    domain : salobj.Domain
+        Optional domain to use for the remotes. If `None` (Default), create a
+        new remote.
+    log : logging.Logger
+        Optional logging class to be used for logging operations. If `None`,
+        creates a new logger. Useful to use in salobj.BaseScript and allow
+        logging in the class use the script logging.
+    intended_usage: int
+        Optional integer that maps to a list of intended operations. This is
+        used to limit the resources allocated by the class by gathering some
+        knowledge about the usage intention. By default allocates all
+        resources.
+
+    Notes
+    -----
+
+    The intended_usage is a bitwise operation which allows users to construct
+    use-cases by summing up usages, e.g.
+
+    `BaseUsages.StateTransition+BaseUsages.MonitorHeartBeat`
```

Comment:
Thank you for offering a way to limit usage.

Note that bit masks are usually combined using `|` instead of `+`, though I guess you get the same result either way, unless the masks have overlapping bits, in which case `|` is definitely the way to go.